### ISSUE TITLE
JDK 15: Illegal reflective access to java.time classes

### ISSUE THREAD

**Issue by eldaragalarov** [google/gson#1807]
JDK 15 produces many Illegal reflective access warnings like following on serializing LocalTime, LocalDateTime, YearMonth, MonthYear and so on:

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.gson.internal.reflect.UnsafeReflectionAccessor (file:/home/eldar/.gradle/caches/modules-2/files-2.1/com.google.code.gson/gson/2.8.6/9180733b7df8542621dc12e21e87557e8c99b8cb/gson-2.8.6.jar) to field java.time.MonthDay.month
WARNING: Please consider reporting this to the maintainers of com.google.gson.internal.reflect.UnsafeReflectionAccessor
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

Code example

```
Gson gson = new GsonBuilder().create();
System.out.println(gson.toJson(LocalDate.now()));
System.out.println(gson.toJson(LocalDateTime.now()));
System.out.println(gson.toJson(YearMonth.now()));
System.out.println(gson.toJson(MonthDay.now()));
```

**Comment by Marcono1234** [comment]
Same as #1059. Gson currently has no built-in support for classes added in Java 7+. It therefore falls back to the reflection based approach for serializing these classes, which is likely not what you want and also makes your code dependent on the implementation details of the JDK (which is why you are seeing these warnings).

So you would have to write your own [`TypeAdapter`s](https://javadoc.io/doc/com.google.code.gson/gson/2.8.6/com.google.gson/com/google/gson/TypeAdapter.html) to get support for these classes or use libraries which add support for them. https://github.com/gkopff/gson-javatime-serialisers might be helpful though I have not used it yet.

**Comment by EldarAgalarov** [comment]
> Same as #1059. Gson currently has no built-in support for classes added in Java 7+. It therefore falls back to the reflection based approach for serializing these classes, which is likely not what you want and also makes your code dependent on the implementation details of the JDK (which is why you are seeing these warnings).
> 
> So you would have to write your own [`TypeAdapter`s](https://javadoc.io/doc/com.google.code.gson/gson/2.8.6/com.google.gson/com/google/gson/TypeAdapter.html) to get support for these classes or use libraries which add support for them. https://github.com/gkopff/gson-javatime-serialisers might be helpful though I have not used it yet.

Yes, as a workaround, I wrote my own adapters like below:

```
public final class LocalDateTimeJsonAdapter implements JsonSerializer<LocalDateTime>, JsonDeserializer<LocalDateTime> {

    public static final LocalDateTimeJsonAdapter INSTANCE = new LocalDateTimeJsonAdapter();


    private LocalDateTimeJsonAdapter() {
    }

    @Override
    public LocalDateTime deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        if (!json.isJsonNull()) {
            return LocalDateTime.parse(json.getAsString(), DateTimeFormatter.ISO_LOCAL_DATE_TIME);
        }
        return null;
    }

    @Override
    public JsonElement serialize(LocalDateTime src, Type typeOfSrc, JsonSerializationContext context) {
        if (src != null) {
            return new JsonPrimitive(src.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
        }
        return JsonNull.INSTANCE;
    }
}
```

**Comment by Marcono1234** [comment]
Note that it is recommended to prefer `TypeAdapter` over `JsonSerializer` and `JsonDeserializer` since it reads the JSON data in a streaming way instead of parsing it as an in-memory representation of `JsonElement`s. The code for it will be pretty similar to your current adapter.

**Comment by EldarAgalarov** [comment]
I will look at it. Thank you!

**Comment by Marcono1234** [comment]
Can probably be closed as duplicate of #1059.
